internal/app/go-filecoin/plumbing/msg/waiter.go

```diff
@@ -149,149 +134,94 @@ func (w *Waiter) waitForMessage(ctx context.Context, ch <-chan interface{}, msgC
-							}
-							return &ChainMessage{msg, blk, recpt}, true, nil
-						}
-					}
-				}
+				return w.receiptForTipset(ctx, raw, msgCid)
 			default:
 				return nil, false, fmt.Errorf("unexpected type in channel: %T", raw)
 			}
 		}
 	}
 }
 
-// receiptFromTipSet finds the receipt for the message with msgCid in the
-// input tipset.  This can differ from the message's receipt as stored in its
-// parent block in the case that the message is in conflict with another
-// message of the tipset.
-func (w *Waiter) receiptFromTipSet(ctx context.Context, msgCid cid.Cid, ts block.TipSet) (*types.MessageReceipt, error) {
-	// Receipts always match block if tipset has only 1 member.
-	var rcpt *types.MessageReceipt
-	if ts.Len() == 1 {
-		b := ts.At(0)
-		// TODO #3194: this should return an error if a receipt doesn't exist.
-		// Right now doing so breaks tests because our test helpers
-		// don't correctly apply messages when making test chains.
-		//
-		j, err := w.msgIndexOfTipSet(ctx, msgCid, ts, make(map[cid.Cid]struct{}))
+func (w *Waiter) receiptForTipset(ctx context.Context, ts block.TipSet, msgCid cid.Cid) (*ChainMessage, bool, error) {
+	tsMessages := make([][]*types.SignedMessage, ts.Len())
+	for i := 0; i < ts.Len(); i++ {
+		blk := ts.At(i)
+		secpMsgs, blsMsgs, err := w.messageProvider.LoadMessages(ctx, blk.Messages)
 		if err != nil {
-			return nil, err
+			return nil, false, err
 		}
 
-		receipts, err := w.messageProvider.LoadReceipts(ctx, b.MessageReceipts)
-		if err != nil {
-			return nil, err
+		cids := make([]cid.Cid, len(blsMsgs)+len(secpMsgs))
+		blkMsgs := make([]*types.SignedMessage, len(blsMsgs)+len(secpMsgs))
+		for j, msg := range blsMsgs {
+			c, err := msg.Cid()
+			if err != nil {
+				return nil, false, err
+			}
+			cids[j] = c
+			blkMsgs[j] = &types.SignedMessage{Message: *msg}
 		}
-		if j < len(receipts) {
-			rcpt = receipts[j]
+		for j, msg := range secpMsgs {
+			c, err := msg.Cid()
+			if err != nil {
+				return nil, false, err
+			}
+			cids[len(blsMsgs)+j] = c
+			blkMsgs[len(blsMsgs)+j] = msg
 		}
-		return rcpt, nil
-	}
+		tsMessages[i] = blkMsgs
 
-	// Apply all the tipset's messages to determine the correct receipts.
-	ids, err := ts.Parents()
-	if err != nil {
-		return nil, err
-	}
-	st, err := w.chainReader.GetTipSetState(ctx, ids)
-	if err != nil {
-		return nil, err
-	}
-
-	tsHeight, err := ts.Height()
-	if err != nil {
-		return nil, err
-	}
-	ancestorHeight := types.NewBlockHeight(tsHeight).Sub(types.NewBlockHeight(consensus.AncestorRoundsNeeded))
-	parentTs, err := w.chainReader.GetTipSet(ids)
-	if err != nil {
-		return nil, err
-	}
-	ancestors, err := chain.GetRecentAncestors(ctx, parentTs, w.chainReader, ancestorHeight)
-	if err != nil {
-		return nil, err
-	}
+		for i, msg := range blkMsgs {
+			if cids[i].Equals(msgCid) {
+				// wrapped cid might be different from given cid
+				targetCid, err := msg.Cid()
```

Comment:
potentially clearer to call this wrappedCid here.

Reply:
Probably a good idea to file an issue sorting the UX out for BLS messages.  Doing it this way requires that all users get the cid of an unsigned message as output from the CLI for BLS and a signed message as output for SECP so that they can correctly wait.